trinity/tools/async_process_runner.py

```diff
@@ -0,0 +1,63 @@
+import asyncio
+import os
+import signal
+from typing import (
+    AsyncIterable,
+    Awaitable,
+    Callable,
+    Iterable,
+)
+
+
+AwaitableGeneratorProducer = Callable[[], Awaitable[bytes]]
+
+
+class AsyncProcessRunner():
+
+    @classmethod
+    async def create_and_run(cls, cmds: Iterable[str], timeout_sec: int=10) -> 'AsyncProcessRunner':
+        runner = cls()
+        await runner.run(cmds, timeout_sec)
+        return runner
+
+    async def run(self, cmds: Iterable[str], timeout_sec: int=10) -> None:
+        proc = await asyncio.create_subprocess_exec(
+            *cmds,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+            # We need this because Trinity spawns multiple processes and we need to take down
+            # the entire group of processes.
+            preexec_fn=os.setsid
+        )
+        self.proc = proc
+        asyncio.ensure_future(self.kill_after_timeout(timeout_sec))
+
+    async def iterate_stdout(self) -> AsyncIterable[str]:
+        async for line in self._iterate_until_empty(self.proc.stdout.readline):
+            yield line
+
+    async def iterate_stderr(self) -> AsyncIterable[str]:
+        async for line in self._iterate_until_empty(self.proc.stderr.readline):
+            yield line
+
+    async def _iterate_until_empty(
+            self,
+            awaitable_bytes_fn: Callable[[], Awaitable[bytes]]) -> AsyncIterable[str]:
+
+        while True:
+            line = await awaitable_bytes_fn()
+            # Leaving this print here allows running pytest with -s and observing the output
+            # There's probably a better way to handle that
+            print(line)
```

Comment:
@pipermerriam @carver This bothers me but I've tried to use a logger and even with `logger.warning` and various kind of pytest arguments I wasn't able to see the output of the integration tests. I wonder if that's related to us logging on `STDERR`?

So, for now I kept the `print` here because it's really useful to be able to run these tests with `pytest -s` and observe the Trinity output. Any ideas are welcome :)

Reply:
Without seeing how you had logging setup, I'm not sure I can be of much help.

Reply:
@pipermerriam I created a new logger like this:

```python
def __init__(self):
    self.logger = logging.getLogger('trinity.tools.async_process_runner.AsyncProcessRunner')

and then used `self.logger.info(...) or even self.logger.warning(...)` instead of the `print`
```

Reply:
Ok, I got rid of this hardcoded `print` inside the `AsyncProcessRunner` and instead allowed passing a`debug_fn` in the `__init__`. That way, at least the `print` got moved into the tests directly.